portfolio/src/main/java/com/google/sps/data/CalendarGroup.java

```diff
@@ -0,0 +1,132 @@
+
+  // These two variables represent the start and end of
+  // possible schedule time. When these times are passed in
+  // to instantiate a CalendarGroup class, these times are
+  // already adjusted based on user's working hour times
+  // and the current timestamp of when the user starts
+  // the scheduling process.
+  public final Instant startTime;
+  public final Instant endTime;
+
+  // Comparator for sorting events by start time
+  private static final Comparator<CalendarEvent> sortByEventStartTimeAscending =
+      Comparator.comparing(CalendarEvent::getStartTime);
+
+  // Comparator for sorting free time ranges by start time
+  private static final Comparator<TimeRange> sortByTimeRangeStartTimeAscending =
+      Comparator.comparing(TimeRange::start);
+
+  /**
+   * @param events: a list of CalendarEvents' that represents the events already scheduled for the
+   *     user.
+   * @param startTime: start of possible scheduling blocks, of type Instant. This is the earlier of
+   *     working hour start time and the current timestamp.
+   * @param endTime: end of possible scheduling blocks, of type Instant.
+   */
+  public CalendarGroup(List<CalendarEvent> events, Instant startTime, Instant endTime) {
+    if (events == null) {
+      throw new IllegalArgumentException("Calendar Group needs a collection of events");
+    }
+    if (startTime == null) {
+      throw new IllegalArgumentException("Calendar Group needs the start of scheduling hours");
+    }
+    if (endTime == null) {
+      throw new IllegalArgumentException("Calendar Group needs the end of scheduling hours");
+    }
+    if (startTime.isAfter(endTime)) {
+      throw new IllegalArgumentException("Start time cannot be after end time");
+    }
+    this.events = events;
+    this.startTime = startTime;
+    this.endTime = endTime;
+    this.allFreeTimeRanges = calculateFreeTimeRanges();
+  }
+
+  /**
+   * Calculates a TimeRange list which represents the periods of time that are empty of events and
+   * lie completely inside the possible scheduling hours. Originally written by tomasalvarez,
+   * modified by hollyyuqizheng.
+   */
+  public List<TimeRange> calculateFreeTimeRanges() {
+    Collections.sort(events, sortByEventStartTimeAscending);
+
+    List<TimeRange> possibleTimes = new ArrayList<TimeRange>();
+
+    // This represents the earliest time that we can schedule a window for the
+    // meeting. As events are processed, this changes to their end times.
+    Instant earliestNonScheduledInstant = startTime;
+    for (CalendarEvent event : events) {
+      // Make sure that there is some time between the events and it is not
+      // later than the person's scheduling hours' ending time.
+      if (event.getStartTime().isAfter(earliestNonScheduledInstant)
+          && !event.getStartTime().isAfter(endTime)) {
+        possibleTimes.add(
+            TimeRange.fromStartEnd(earliestNonScheduledInstant, event.getStartTime()));
+      }
```

Comment:
Make sure we include the code that's missing here and a comment: 
`// This is so events outside the time period are not processed.`
`else { continue;}`

Reply:
On second thought, I don't think we should include this continue statement. For example, if the earliestNonScheduledInstant is at 9:00, and we have an event from 8:30 to 9:30, we do not add any new time range into possibleTimes, but we need to shift the earliestNonScheduledInstant to 9:30 now, which is handled in the if-statement that follows, so I don't think we can skip this if statement?